DigitalOcean 1-click Droplet and Kubernetes getting started guides (#8773)
gravitational/teleport: docs/pages/kubernetes-access/helm/guides/digitalocean.mdx

---
title: Get started with Teleport on DigitalOcean Managed Kubernetes cluster.

> description: Learn how to get started with Teleport on DigitalOcean Managed Kubernetes cluster

Reviewer: Will update the description. I think descriptions are for SEO.

---

> This guide will show you how to get started with Teleport on DigitalOcean managed Kubernetes cluster.  Specifically, we will 

Reviewer: Updated "DigitalOcean managed Kubernetes cluster" to "DigitalOcean Kubernetes". Regarding the list, I think it should be numbered list because it presents the primary steps for the tutorial.

1. Create a DigitalOcean Kubernetes cluster, 
2. Install Teleport in this cluster with helm, 
3. Create a Teleport user, create a Kubernetes role and assign a Teleport user with a Kubernetes role.
4. Log in to `tsh` client and access the Kubernetes cluster. 

## Prerequisites
1. DigitalOcean account.
2. Your workstation setup with [Kubectl](https://kubernetes.io/docs/tasks/tools/), [Helm](https://helm.sh/docs/intro/install/), [Doctl](https://docs.digitalocean.com/reference/doctl/how-to/install/), and Teleport [tsh](https://goteleport.com/docs/installation/) client.

<Admonition
  type="note"
  title="Note"
>

> Throughout this tutorial, we are using a domain name `tele.teleporters.dev` as a `clusterName`. By default, Teleport auth and proxy server's address are also configured with the value of a `clusterName`. Therefore, make sure to replace this domain with a domain name you manage.  

Reviewer: Removing this section

</Admonition>

## Step 1/4. Setup DigitalOcean Kubernetes Cluster 
Create a new [DigitalOcean Kubernetes Cluster](https://cloud.digitalocean.com/kubernetes/clusters/)
![Create DigitalOcean Kubernetes Cluster](../../../../img/helm/digitalocean/create-k8s.png)

> While the Kubernetes cluster is being provisioned, set up your kubectl client with reference to getting started guide available in DigitalOcean:

Reviewer: There is no separate page link to add here. The Get Started guide renders in the page with specific values to the newly created Kubernetes cluster. The screenshot below shows this page.

```code

$ kubectl --namespace=teleport-cluster get services
NAME               TYPE           CLUSTER-IP      EXTERNAL-IP       PORT(S)          AGE
# teleport-cluster   LoadBalancer   10.245.163.12   144.126.252.144   443:31959/TCP,3023:30525/TCP,3026:30079/TCP,3024:32437/TCP   19m

```

Once you can see the value for `EXTERNAL-IP`(public IP address), update your DNS record such that the clusterName's A record points to this IP address. For example `144.126.252.144` is the public IP in the above case.

## Step 3/4. Create Teleport User
Now we create a Teleport user by executing the `tctl` command with `kubectl`.

```code
$ kubectl --namespace teleport-cluster exec deploy/teleport-cluster -- tctl users add tadmin --roles=access,editor --logins=root,ubuntu
# User "tadmin" has been created but requires a password. Share this URL with the user to complete user setup, link is valid for 1h:
# https://tele.teleporters.dev:443/web/invite/9b8c8ccb53d0aab612487d279d3d9831

# NOTE: Make sure tele.teleporters.dev:443 points at a Teleport proxy which users can access.
```

> Copy the link shown after executing the above command and open the link in a web browser to complete user registration process (the link is `https://tele.teleporters.dev:443/web/invite/9b8c8ccb53d0aab612487d279d3d9831` in the above case).

Reviewer 2: "user registration process" -> "the user registration process"

![Setup User](../../../../img/helm/digitalocean/setup-user.png)

> After you complete the registration process by setting up a password and enrolling in two-factor authentication, you will be logged in to Teleport WebUI. 

Reviewer 2: "Teleport WebUI" -> "the Teleport web UI"

In this step, we created a user **tadmin** with roles `access, edit`. These are the default roles available in Teleport. However, to allow this user to access the Kubernetes cluster, we will need to assign **tadmin** a role authorized to access the Kubernetes cluster. So first, let's create a role named **member** with the Kubernetes privilege `system:master`.

> **Create a new Kubernetes Role name "member":**

Reviewer 2: - Should this be a heading rather than bolded text?\
- "name" -> "named"

> Create a file name `member.yaml` with the following spec:

Reviewer 2: "name" -> "named"

```code

> $ vi member.yaml

Reviewer 2: Since we already told the reader to create the file, I don't think we need the `vi` command here. Maybe just add a `yaml` snippet instead of a `code` one?

kind: role
version: v4
metadata:
  name: member
spec:
  allow:
    kubernetes_groups: ["system:masters"]

```
<Admonition
  type="warning"
  title="Warning"
>

> `system:masters` is a Kubernetes built-in role that is similar to `sudo` permissions in Linux systems. It is used here for demonstration purposes and should be sparingly used in production.

Reviewer 2: > should be sparingly used in production

Maybe make this more strongly worded, e.g., "should be used in production only when necessary."

</Admonition>

Next, create this role in Kubernetes with the command:

```code

$ POD=$(kubectl --namespace=teleport-cluster get pod -l app=teleport-cluster -o jsonpath='{.items[0].metadata.name}')
$ kubectl --namespace=teleport-cluster exec -i ${POD?} -- tctl create -f < member.yaml
```

> **Assign "member" role to user "tadmin":**

Reviewer 2: - I think there should be a "the" before "member"
- I think this should be a heading rather than bold text

Now we will assign Teleport user **tadmin** with this role. The example below shows a process using Teleport WebUI:
First, lets select user edit menu:
![Edit User](../../../../img/helm/digitalocean/edit-user.png)

> Second, update user role to assign **member** role:

Reviewer 2: I think this should be:

> Second, update the `tadmin` user to assign the **member** role:

![Update Role](../../../../img/helm/digitalocean/update-role.png)

> We've update user **tadmin** with a role **member** whish is allowed to access Kubernetes cluster with privilege `system:master`.

Reviewer 2: I think this should be,

> We've updated the user **tadmin** to have the **member** role, which is allowed to access our Kubernetes cluster with the privilege `system:master`.

> ## Step 4/4. Connect to Kubernetes cluster using Teleport

Reviewer 2: I think this should be,

> Connect to your Kubernetes cluster using Teleport

The following steps show how to access the Kubernetes cluster using `tsh`.

**Login to Teleport:**

> First, login to Teleport using `tsh`.

Reviewer 2: I think "login to" should be "log into"

```code 

> $ tsh login --proxy=tele.teleporters.dev:443 --auth=local --user=tadmin(we created this user earlier)

Reviewer 2: Rather than the `(we created this user earlier)` parenthetical, I think we should use a code comment

Enter password for Teleport user tadmin:
Enter your OTP token:
540255
> Profile URL:        https://tele.teleporters.dev:443
  Logged in as:       tadmin
  Cluster:            tele.teleporters.dev
  Roles:              access, editor
  Logins:             root, ubuntu
  Kubernetes:         enabled
  Valid until:        2021-10-27 06:37:15 +0000 UTC [valid for 12h0m0s]
  Extensions:         permit-agent-forwarding, permit-port-forwarding, permit-pty
```

> To write kubectl configuration to a separate file instead of having your global kubectl configuration modified, run the following command:

Reviewer 2: "To write kubectl" -> "To write your kubectl"

```code
$ export KUBECONFIG=${HOME?}/teleport-kubeconfig.yaml
```

**Select the Kubernetes cluster:**
```code
$ tsh kube login tele.teleporters.dev
Logged into kubernetes cluster "tele.teleporters.dev"
```

**Connect to the Kubernetes cluster:**
```code
$ kubectl --namespace=teleport-cluster get pods
NAME                                READY   STATUS    RESTARTS   AGE
teleport-cluster-6cc679b6f6-7xr5h   1/1     Running   0          14h
```

> Voila! User **tadmin** was able to list pod in DigitalOcean Kubernetes cluster. 

Reviewer 2: I think this should be, 

> User **tadmin** was able to list the pods in their DigitalOcean Kubernetes cluster.

> Teleport keeps an audit log of access made by users. The below image shows a screenshot of the Teleport WebUI audit log page showing user **tadmin** access to the Kubernetes cluster.

Reviewer 2: I think this should be,

> Teleport keeps an audit log of access to a Kubernetes cluster. In the screenshot below, the Teleport audit log shows that the user **tadmin** has logged into the cluster.

</Figure>

## Step 2/4. Install Teleport 

Add the Teleport repository to Helm.

```code
$ helm repo add teleport https://charts.releases.teleport.dev
```

Install Teleport in your Kubernetes cluster using the `teleport-cluster` Helm chart.
```code
$ CLUSTERNAME=tele.teleporters.dev # replace with your preferred domain name
$ EMAIL_ADDR=dodemo@goteleport.com # replace with your email
$ helm install teleport-cluster teleport/teleport-cluster --create-namespace --namespace=teleport-cluster \
  --set clusterName=$CLUSTERNAME \
  --set acme=true \
  --set acmeEmail=$EMAIL_ADDR
NAME: teleport-cluster
LAST DEPLOYED: Tue Oct 26 17:01:21 2021
NAMESPACE: teleport-cluster
STATUS: deployed
REVISION: 1
TEST SUITE: None
```

### Update DNS for `clusterName`
First, get the external IP (from the `EXTERNAL-IP` field) for the Kubernetes cluster. 
```code

$ kubectl --namespace=teleport-cluster get services
NAME               TYPE           CLUSTER-IP      EXTERNAL-IP       PORT(S)          AGE
# teleport-cluster   LoadBalancer   10.245.163.12   192.168.200.200   443:31959/TCP,3023:30525/TCP,3026:30079/TCP,3024:32437/TCP   19m

```

> Once you get the value for external IP (it may take a few minutes for this field to be populated), update your DNS record such that the clusterName's A record points to this IP address. For example `192.168.200.200` is the external IP in the above case.

Reviewer 2: ```suggestion
Once you get the value for the external IP (it may take a few minutes for this field to be populated), update your DNS record such that the clusterName's A record points to this IP address. For example `192.168.200.200` is the external IP in the above case.
```

<Figure align="left" bordered caption="Configure DNS">
  ![Configure DNS](../../../../img/helm/digitalocean/fqdn.png)
</Figure>

> ## Step 3/4. Create and set up Teleport user.

Reviewer 2: ```suggestion
## Step 3/4. Create and set up Teleport user
```
for consistency

Now we create a Teleport user by executing the `tctl` command with `kubectl`.

```code

> $ kubectl --namespace teleport-cluster exec deploy/teleport-cluster -- tctl users add tadmin --roles=access,editor --logins=root,ubuntu

Reviewer 3: ```suggestion
$ kubectl --namespace teleport-cluster exec deploy/teleport-cluster -- tctl users add tadmin --roles=access,editor --logins=ubuntu  
```

Dropping root as a best practice.

# User "tadmin" has been created but requires a password. Share this URL with the user to complete user setup, link is valid for 1h:

> # https://tele.teleporters.dev:443/web/invite/<invite-token>

Reviewer 3: ```suggestion
# https://tele.example.com:443/web/invite/<invite-token>
```

# NOTE: Make sure tele.teleporters.dev:443 points at a Teleport proxy which users can access.
```

Copy the link shown after executing the above command and open the link in a web browser to complete the user registration process (the link is `https://tele.teleporters.dev:443/web/invite/<invite-token>` in the above case).
<Figure align="left" bordered caption="Setup user">
  ![Setup user](../../../../img/helm/digitalocean/setup-user.png)
</Figure>

After you complete the registration process by setting up a password and enrolling in two-factor authentication, you will be logged in to Teleport Web UI. 

In this step, we created a user **tadmin** with roles `access, edit`. These are the default roles available in Teleport. However, to allow this user to access the Kubernetes cluster, we will need to assign **tadmin** a role authorized to access the Kubernetes cluster. So first, let's create a role named **member** with the Kubernetes privilege `system:master`.

> ### Create a new Kubernetes Role named "member".

Reviewer 2: ```suggestion
### Create a new Kubernetes Role named "member"
```

Create a file named `member.yaml` with the following spec:
```yaml
kind: role
version: v4
metadata:
  name: member
spec:
  allow:
    kubernetes_groups: ["system:masters"]

> 

Reviewer 3: ```suggestion
```

```
<Admonition
  type="warning"
  title="Warning"
>
`system:masters` is a Kubernetes built-in role that is similar to `sudo` permissions in Linux systems. It is used here for demonstration purposes and should be used in production only when necessary.
</Admonition>

Next, create this role in Kubernetes with the command:

```code

$ POD=$(kubectl --namespace=teleport-cluster get pod -l app=teleport-cluster -o jsonpath='{.items[0].metadata.name}')
$ kubectl --namespace=teleport-cluster exec -i ${POD?} -- tctl create -f < member.yaml
```

### Assign the "member" role to user "tadmin"

Now we will assign Teleport user **tadmin** with this role. The example below shows a process using Teleport Web UI:
First, lets select user edit menu:
<Figure align="left" bordered caption="Edit user">
  ![Edit user](../../../../img/helm/digitalocean/edit-user.png)
</Figure>

> Second, update the **tadmin** user role to assign **member** role:

Reviewer 2: ```suggestion
Second, update the **tadmin** user role to assign the **member** role:
```

<Figure align="left" bordered caption="Update role">
  ![Update role](../../../../img/helm/digitalocean/update-role.png)
</Figure>

We've updated the user **tadmin** to have the **member** role, whish is allowed to access Kubernetes cluster with privilege `system:master`.

## Step 4/4. Access your Kubernetes cluster using Teleport

The following steps show how to access the Kubernetes cluster using `tsh`.

> ### Login to Teleport.

Reviewer 2: ```suggestion
### Log into Teleport
```

First, authenticate to Teleport using `tsh` as the **tadmin** user we created in the third step. 

<Admonition
  type="note"
  title="Note"
>
The following process will override your global kubectl configuration file. To write your kubectl configuration to a separate file instead of having your global kubectl configuration modified, run the following command:

```code
$ export KUBECONFIG=${HOME?}/teleport-kubeconfig.yaml
```

</Admonition>

```code 
$ tsh login --proxy=tele.teleporters.dev:443 --auth=local --user=tadmin 
Enter password for Teleport user tadmin:
Enter your OTP token:
540255
> Profile URL:        https://tele.teleporters.dev:443
  Logged in as:       tadmin
  Cluster:            tele.teleporters.dev

>   Roles:              access, editor

Reviewer 3: ```suggestion
  Roles:              access, editor, member
```

  Logins:             root, ubuntu
  Kubernetes:         enabled
  Valid until:        2021-10-27 06:37:15 +0000 UTC [valid for 12h0m0s]
  Extensions:         permit-agent-forwarding, permit-port-forwarding, permit-pty
```

> ### Select the Kubernetes cluster.

Reviewer 2: ```suggestion
### Select the Kubernetes cluster
```

```code

> $ tsh kube login tele.teleporters.dev

Reviewer 3: ```suggestion
$ tsh kube login tele.example.com
```

> Logged into kubernetes cluster "tele.teleporters.dev"

Reviewer 3: ```suggestion
Logged into kubernetes cluster "tele.example.com"
```

```

> ### Connect to the Kubernetes cluster.

Reviewer 2: ```suggestion
### Connect to the Kubernetes cluster
```

---

> title: Get started with Teleport on DigitalOcean Kubernetes.

Reviewer 2: ```suggestion
title: Get started with Teleport on DigitalOcean Kubernetes
```

> description: How to get started with Teleport on DigitalOcean Kubernetes?

Reviewer 2: ```suggestion
description: How to get started with Teleport on DigitalOcean Kubernetes
```

---

This guide will show you how to get started with Teleport on DigitalOcean Kubernetes. 

## Prerequisites
- DigitalOcean account.
- Your workstation configured with [kubectl](https://kubernetes.io/docs/tasks/tools/), [Helm](https://helm.sh/docs/intro/install/), [doctl](https://docs.digitalocean.com/reference/doctl/how-to/install/), and the Teleport [tsh](https://goteleport.com/docs/installation/) client.

## Step 1/4. Create a DigitalOcean Kubernetes cluster
Create a new [DigitalOcean Kubernetes Cluster](https://cloud.digitalocean.com/kubernetes/clusters/)
<Figure align="left" bordered caption="Create DigitalOcean Kubernetes cluster">
  ![Create DigitalOcean Kubernetes cluster](../../../../img/helm/digitalocean/create-k8s.png)
</Figure>

<br />

> While the Kubernetes cluster is being provisioned, follow the "Get started" guide as shown below:

Reviewer 2: ```suggestion
While the Kubernetes cluster is being provisioned, follow the "Getting Started" guide as shown below:
```
so this matches the screenshot

<Figure align="left" bordered caption="Set up DigitalOcean Kubernetes client">
  ![Set up DigitalOcean Kubernetes client](../../../../img/helm/digitalocean/setup-k8s.png)
</Figure>

## Step 2/4. Install Teleport 

> Add Teleport repository to Helm.

Reviewer 2: ```suggestion
Add the Teleport repository to Helm.
```

```code
$ helm repo add teleport https://charts.releases.teleport.dev
```

Install Teleport in your Kubernetes cluster using the `teleport-cluster` Helm chart.
```code

> $ CLUSTERNAME=tele.teleporters.dev(replace with your preferred domain name)

Reviewer 2: ```suggestion
$ CLUSTERNAME=tele.teleporters.dev # replace with your preferred domain name
```

> $ EMAIL_ADDR=dodemo@goteleport.com(replace with your email)

Reviewer 2: ```suggestion
$ EMAIL_ADDR=dodemo@goteleport.com # replace with your email
```

# https://tele.teleporters.dev:443/web/invite/<invite-token>

# NOTE: Make sure tele.teleporters.dev:443 points at a Teleport proxy which users can access.
```

Copy the link shown after executing the above command and open the link in a web browser to complete the user registration process (the link is `https://tele.teleporters.dev:443/web/invite/<invite-token>` in the above case).
<Figure align="left" bordered caption="Setup user">
  ![Setup user](../../../../img/helm/digitalocean/setup-user.png)
</Figure>

After you complete the registration process by setting up a password and enrolling in two-factor authentication, you will be logged in to Teleport Web UI. 

In this step, we created a user **tadmin** with roles `access, edit`. These are the default roles available in Teleport. However, to allow this user to access the Kubernetes cluster, we will need to assign **tadmin** a role authorized to access the Kubernetes cluster. So first, let's create a role named **member** with the Kubernetes privilege `system:master`.

### Create a new Kubernetes Role named "member".

Create a file named `member.yaml` with the following spec:

> ```code

Reviewer 2: ```suggestion
```yaml
```

$ kubectl --namespace=teleport-cluster exec -i ${POD?} -- tctl create -f < member.yaml
```

### Assign the "member" role to user "tadmin"

Now we will assign Teleport user **tadmin** with this role. The example below shows a process using Teleport Web UI:
First, lets select user edit menu:
<Figure align="left" bordered caption="Edit user">
  ![Edit user](../../../../img/helm/digitalocean/edit-user.png)
</Figure>

> Second, update **tadmin** user role to assign **member** role:

Reviewer 2: ```suggestion
Second, update the **tadmin** user role to assign **member** role:
```

<Figure align="left" bordered caption="Update role">
  ![Update role](../../../../img/helm/digitalocean/update-role.png)
</Figure>

> We've updated the user **tadmin** to have the **member** role, whish is allowed to access Kubernetes cluster with privilege `system:master`.

Reviewer 2: ```suggestion
We've updated the user **tadmin** to have the **member** role, which is allowed to access a Kubernetes cluster with privilege `system:master`.
```

## Step 4/4. Access your Kubernetes cluster using Teleport

The following steps show how to access the Kubernetes cluster using `tsh`.

### Login to Teleport.

> First, authenticate to Teleport using `tsh` as a **tadmin** user which we created earlier in the third step. 

Reviewer 2: ```suggestion
First, authenticate to Teleport using `tsh` as the **tadmin** user we created in the third step. 
```
